foxai/explainer/computer_vision/algorithm/xrai.py

```diff
@@ -0,0 +1,687 @@
+
+        Returns:
+            Batch tensor.
+        """
+        # add 2 artificial dimensions
+        # first for artificial batch size which will be stacked
+        # second for artificial 1D channel
+        batch_attributes: List[np.ndarray] = []
+        if isinstance(result_list[0], list):
+            for inner_list in result_list:
+                batch_attributes.append(np.stack(list(inner_list)))
+        else:
+            batch_attributes = [np.expand_dims(val, 0) for val in result_list]
+
+        return np.stack(batch_attributes)
+
+    @staticmethod
+    def _xrai(
+        attributes: np.ndarray,
+        segment_list: List[List[np.ndarray]],
+        gain_fun: Callable[
+            [np.ndarray, np.ndarray, Optional[np.ndarray]], float
+        ] = _gain_density,
+        area_perc_th: float = 1.0,
+        min_pixel_diff: int = 50,
+        integer_segments: bool = True,
+    ) -> Tuple[np.ndarray, Union[np.ndarray, List[np.ndarray]]]:
+        """Run XRAI saliency given attributions and segments.
+
+        Args:
+            attributes: Source attributions for XRAI.
+            segment_list: Input segments as a list of boolean masks. XRAI uses these to
+                compute attribution sums. Fist level list is for batch dimension.
+            gain_fun: The function that computes XRAI area attribution from source
+                attributions. Defaults to `_gain_density`, which calculates the
+                density of attributions in a mask.
+            area_perc_th: The saliency map is computed to cover `area_perc_th` of
+                the image. Lower values will run faster, but produce
+                uncomputed areas in the image that will be filled to satisfy
+                completeness. Defaults to `1.0`.
+            min_pixel_diff: Do not consider masks that have difference less than
+                this number compared to the current mask. Set it to 1
+                to remove masks that completely overlap with the
+                current mask.
+            integer_segments: If set to `True`, the XRAI segments are returned as an
+                integer array with the same dimensions as the input (excluding color
+                channels). The elements of the array are set to values from the
+                `[1, N]` range, where `1` is the most important segment and `N` is the
+                least important segment. If integer_segments is set to `False`,
+                the segments are returned as a boolean array, where the first
+                dimension has size `N`. The `[0, ...]` mask is the most important and the
+                `[N-1, ...]` mask is the least important. Defaults to `True`.
+
+        Returns:
+            Saliency heatmap `np.ndarray` and masks or an integer image with
+                area ranks depending on the parameter integer_segments.
+        """
+        attributes_list: List[np.ndarray] = []
```

Comment:
AttributionsType?

Reply:
`AttributionsType` is defined as `torch.Tensor`. All types have to be coherent in public methods/API but internally it makes more sense be less restrictive.